src/tree/values/FormattingElement.js

```diff
@@ -0,0 +1,45 @@
+/**
+ * Represents formatting elements (e.g. link, image, bold text) within a document.
+ */
+class FormattingElement {
+	/**
+	 * Represents a formatting element (e.g. link, image, bold text) within a document.
+	 *
+	 * @param {string} type          		The type of this element ("link", "image", "bold", etc.).
+	 * @param {number} start        		The start position of this element within the text.
+	 * @param {number} end               	The end position of the element within the text.
+	 * @param {Object} [attributes=null] 	The attributes (as key-value pairs, e.g. "href='...'" => { href: '...' } ).
+	 */
+	constructor( type, start, end, attributes = null ) {
+		this.type = type;
+		this.attributes = attributes;
+		this.start = start;
+		this.end = end;
+
+		// Swap end and start positions when end is smaller.
+		if ( this.end < this.start ) {
+			this._swapStartAndEnd();
+			console.warn( `End position smaller than start of '${type}' element. They have been swapped.` );
+		}
+
+		// Set start position to zero when smaller than zero.
+		if ( this.start < 0 ) {
```

Comment:
You are right. I thought that, when `endIndex` < 0, it would always be swapped with `startIndex` by the code before. In this way, I thought, `endIndex` would never be smaller than zero. 
However, both `startIndex` and `endIndex` could be smaller than zero and not be swapped (e.g. when `startIndex = -8` and `endIndex = -6`).

I did change the code to check `startIndex` and `endIndex` separately, since we do not want the `endIndex` to be zero when only the `startIndex` is smaller than zero (I think?).